range-local intent resolution

tbg opened the pull request:
this passes all tests except for one (which I've disabled; maybe
@bdarnell has an idea?). Basically carried out a lot of the work announced in
RFC #1873:
- coordinator sends intents along with EndTransaction
- range resolves all it possibly can in the EndTransaction batch, and hands
  everything else off to an async resolver.
- "skipped intents" passed up by the range_command functions are now executed
  even if the command returns an error. This is required for EndTransaction
  which finds out that it's been aborted.

I addressed all of the TODOs I wanted to deal with before merging. The
transaction record cleanup mechanism isn't there yet: it's the natural candidate
for the next PR. I think this is ready to merge after adding some more tests,
so it's ready for a closer look.

bdarnell commented on proto/data.go:
```diff
@@ -96,7 +96,14 @@ func (k Key) Next() Key {
+// IsPrev is a more efficient version of k.Next().Equal(m).
+func (k Key) IsPrev(m Key) bool {
+	l := len(m) - 1
+	return l == len(k) && m[l] == 0
```
This isn't enough. We still need to check that `k` and `m[:l]` are equal (consider `foo` and `bar\x00`).

tschottdorf replied:
geez. Good point.

bdarnell commented on storage/client_raft_test.go:
```diff
@@ -863,5 +864,6 @@ func TestRaftAfterRemoveRange(t *testing.T) {
 		t.Fatal(err)
 	}
 	// Execute another replica change to ensure that MultiRaft has processed the heartbeat just sent.
+	// TODO(tschottdorf): without this line, test passes. Gotta investigate.
```
What's the failure? We're tracking a race in this test in #1911.

tschottdorf replied:
it just never finds the new replica.

`V(1)`: https://gist.github.com/tschottdorf/3380546a93f8851e5bbd
`V(5)`: https://gist.github.com/tschottdorf/be8122421d3541e25aa1

for reference, here's the `V(1)` of current master.
https://gist.github.com/tschottdorf/dc7f5cae1de7d99bf41c

bdarnell replied:
OK, so it looks like a timing issue that was exposed by the fact that `EndTransaction` now takes longer. It has to do with coalesced heartbeats. After range 2 is removed, node 2 believes that it has no ranges in common with node 1, so it doesn't respond to heartbeats. We have a special case to send heartbeats when the receiving node doesn't know anything about the sending node, but nothing for when the recipient doesn't think there are any ranges in common. I think we just need to remove the `cnt` check from `fanoutHeartbeat` and reply unconditionally.

bdarnell commented on storage/range_command.go:
```diff
@@ -229,43 +229,66 @@ func (r *Range) Scan(batch engine.Engine, args proto.ScanRequest) (proto.ScanRes
-func (r *Range) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args proto.EndTransactionRequest) (proto.EndTransactionResponse, error) {
+func (r *Range) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args proto.EndTransactionRequest) (proto.EndTransactionResponse, []proto.Intent, error) {
 	var reply proto.EndTransactionResponse
+	abort := func(err error) (proto.EndTransactionResponse, []proto.Intent, error) {
```
This is an odd pattern; I'd rather just `return reply, nil, util.Errorf(...)` at each call site.

tschottdorf replied:
ok, done.

bdarnell commented:
LGTM so far.

mrtracy commented:
Very well commented; everything appears to do what it claims, by my eye. LGTM

tamird commented on storage/store.go:
```diff
@@ -1398,40 +1422,90 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
+		// need-to-resolve-some-intents-or-a-split-may-never-finish thing
+		// going on on during node shutdown.
+		if !r.ContainsKey(intent.Key) {
+			b := &client.Batch{}
```
i'm confused, isn't this using a separate batch for each intent?

tschottdorf replied:
yes, see the comment.

tamird commented on storage/store.go:
```diff
@@ -1398,40 +1422,90 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
+			// a no-op if all are local.
+			bReply := &proto.InternalBatchResponse{}
+			b.InternalAddCall(proto.Call{Args: bArgs, Reply: bReply})
+			if resolveErr := r.rm.DB().Run(b); resolveErr != nil {
```
i think you can name this `err` now that the name is not conflicting with anything

tamird replied:
also i think this needs to be a task on the stopper

tschottdorf replied:
ditto.

tamird replied:
ditto as in "see the comment"? i don't follow

tschottdorf commented on storage/store.go:
```diff
@@ -1407,40 +1431,91 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
+		// TODO(tschottdorf): Also no support for key-range intents yet. We
+		// need to support those because the EndTransaction path will throw
+		// them our way. And we need a test that does exactly that.
+		// TODO(tschottdorf): currently synchronous. See if we still have that
```
ping @tamird re: comment below

tamird replied:
even if it's synchronous, it needs to be tied to the stopper. Is this being invoked inside a `RunWorker`?

tschottdorf replied:
it's not important now. This is provisional code until the TODOs above are addressed. We'll want to do this async eventually.

tamird replied:
I think you're asking for a deadlock in a test if you don't `RunTask` here, but ok

tbg committed:
range-local intent resolution

this passes all tests except for one (which I've disabled; maybe
@bdarnell has an idea?). Basically carried out a lot of the work announced in
RFC #1873:

- coordinator sends intents along with EndTransaction
- range resolves all it possibly can in the EndTransaction batch, and hands
  everything else off to an async resolver.
- "skipped intents" passed up by the range_command functions are now executed
  even if the command returns an error. This is required for EndTransaction
  which finds out that it's been aborted.

There are a lot of TODOs documented in the code, and I intend to address all
of those that are relevant for correctness. Some parts of the RFC are also
not yet implemented (gc'ing Txn entries), but the road should be very clear
and I'll add those before merging.

I did want to get this out early because it's a fairly complex refactoring
and will only get bigger from here.

tbg committed:
allow some txns in non-txn batch

... if they are "intent-less" operations and the batch
itself is non-transactional.

this allows sending InternalResolveIntent requests for
multiple transactions' intents in (*Replica).resolveIntents().

tbg committed:
add intent key range handling at Replica level

this addresses most of the comments in `(*Replica).resolveIntents`.
tests for this functionality are next.

tbg committed:
test for local intent resolution

this fairly high-level test makes sure that only non-local intents are
dispatched for resolution via a new RPC. It also documents potential
for future optimizations.
also added a test for verifying range-local key ranges and edited
some comments.

tschottdorf commented:
PTAL, just rebased and added another test for the intent resolution.
The test is still skipped (or should it be fixed now? @bdarnell), I'll open an issue to fix it up if we don't have one yet.

bdarnell commented:
The test still needs to be fixed; go ahead and open another issue.

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -587,22 +545,31 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
 	aHeader.User = bHeader.User
 	aHeader.UserPriority = bHeader.UserPriority
-	aHeader.Txn = bHeader.Txn
+	// Only allow individual transactions in a batch if
```
s/individual transactions/transactions on the individual requests of a batch/

bdarnell replied:
From discussion with @spencerkimball: It's weird that `ResolveIntent` even uses the header transaction, since it is not really a part of that transaction. Instead of allowing requests from multiple transactions to appear in the same batch, it's probably better to just add a Transaction field to ResolveIntentRequest so it doesn't use the header transaction at all (and then we can preserve the invariant that all requests in a batch have the same transaction)

tschottdorf replied:
done. I had the same thoughts about header usage by Resolve and Push (and even carried it out in unpushed code) but decided it's worth its own PR. I'll file an issue.

tschottdorf replied:
#1988

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -484,18 +446,16 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
 						lastUpdateNanos:  tc.clock.PhysicalNow(),
 						timeoutDuration:  tc.clientTimeout,
-						txnEnd:           make(chan []proto.Key, 1),
+						txnEnd:           make(chan []proto.Key),
```
Doesn't look like we're using a value here any more, so `txnEnd` can be changed to a `chan struct{}`.

tschottdorf replied:
done.

bdarnell commented on storage/replica.go:
```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+			}
+		}
+
+		// If the intent isn't (completely) local, we'll need to send an external request.
```
Batching is (or soon will be) worthwhile even for the local range. Maybe this should build up two batches, one local and one remote.

spencerkimball replied:
:+1:

tschottdorf replied:
It's not supported yet. Let's do it when it is.

bdarnell commented on storage/replica.go:
```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+			continue
+		}
+
+		// If it is local, it goes directly into Raft.
```
How much does this split save us? If it's going through raft it already involves non-local communication; it would be simpler to just treat everything as if it were remote.

spencerkimball replied:
Well thing is we actually want to make sure that local intents are resolved synchronously. If we group these all into a single batch and invoke through non-local dist-sender, we'd need to wait for both the fast local intents and the slow non-local intents before finishing the txn.

bdarnell replied:
We're not resolving them synchronously - we're waiting for the commands to be proposed to raft, and then moving on before they are committed (at least if the comments are still accurate). Some local intents are resolved completely synchronously in `EndTransaction`, though. The ones that make it here are just the ones that need some other transaction to be pushed before they can be resolved.

spencerkimball replied:
Scratch this. I see that this is only used from the `Store.resolveWriteIntentError` execution path.

tschottdorf replied:
> How much does this split save us?

good question. The optimization here is to put things into Raft, which means almost certainly that they'll commit before our retrying client gets its next attempt in. So it definitely saves us from having to wait for Raft, which is a couple of round-trips.
I've added a TODO checking this for premature optimization. I imagine we'll refactor this part anyways when we have range-local batch support.

bdarnell commented on storage/replica.go:
```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+
+	// Wait until all the local `ResolveIntent`s have been submitted to raft.
+	// No-op if all were external.
+	wg.Wait()
```
Why do we need to wait for local intents to be submitted to raft, but we don't wait for anything about non-local intents? Is this the real reason for the local vs remote split? If so, mention it in the comments instead of calling the local handling a "fast-path".

spencerkimball replied:
We wait for the local intents to handle the case of ensuring that all range descriptor intents (for example) are cleaned up prior to finishing the commit for a split--can't leave these things dangling.

spencerkimball replied:
Again, scratch my previous comment, which presupposed this method was being called from `EndTransaction`.

tschottdorf replied:
local intents can be conflict resolution with the client waiting, so we want to wait for the minimum amount of time that guarantees that after returning to the client, they'll not hit the intent again. That's why we wait for propose and then return. BTW, nothing has changed for those semantics.

with non-local intents, we know there's no client waiting, so there's no point in blocking the caller. That's the new part introduced.

I'm updating the comments to explain that better.

tschottdorf replied:
> // resolveIntents resolves the given intents. For those which are local to the
// range, we submit directly to the range-local Raft instance; the call returns
// as soon as all resolve commands have been **proposed** (not executed). This
// ensures that if a waiting client retries immediately after conflict
// resolution, it will not hit the same intents again. All non-local intents
// are resolved asynchronously in a batch.

bdarnell commented:
LGTM

spencerkimball commented on storage/replica.go:
```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+		action := func() {
+			// Trace this under the ID of the intent owner.
+			ctx := tracer.ToCtx(ctx, r.rm.Tracer().NewTrace(resolveArgs.Header().Txn))
+			if _, err := r.addWriteCmd(ctx, resolveArgs, &wg); err != nil && log.V(0) {
```
This doesn't seem either necessary or desirable. Why not just call engine.ResolveIntent directly? Calling addWriteCmd unnecessarily threads this through raft again. Keep in mind that all replicas already have the same set of intents to resolve!

bdarnell replied:
This part isn't in the raft command processing - it's called from the Store that originally proposed the EndTransactionCommand after it has been applied.

tschottdorf replied:
I don't understand this comment.

bdarnell replied:
I think I was replying to a comment from @spencerkimball which he has since deleted.

spencerkimball commented on storage/replica.go:
```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
 	return err
 }
 
+// resolveIntents resolves the given intents. Those which are local to the range
```
Why is this broken out? This new method seems more complicated than what was previously in `Store.resolveWriteIntentError()`. We're unlikely to need a slow and fast path bifurcation from write intents encountered executing individual commands.

tschottdorf replied:
check out the comment in `resolveWriteIntentError`. Basically local path = local intents with clients waiting, "slow" path = remote intents that can be done completely async. The latter occurs only now because `EndTransaction` returns its remote intents as skipped intents via existing mechanisms (which seems wise because the alternative is calling directly from very low-level code to very high-level code). If that code isn't here, it's going to be somewhere else, and having a method that you give a bunch of intents to and it does the right thing seems good, though I imagine we'll refactor it a bit over time. BTW, your recent GC intent stuff could probably just reuse parts here instead of patching together its own requests.

spencerkimball commented on storage/replica_command.go:
```diff
@@ -294,27 +312,67 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
-	if err := engine.MVCCPutProto(batch, ms, key, proto.ZeroTimestamp, nil, reply.Txn); err != nil {
-		return reply, err
+	// Resolve any explicit intents. All that are local to this range get
+	// resolved synchronously in the same batch. The remainder is collected
```
s/is/are/

tschottdorf replied:
done.

spencerkimball commented on storage/replica_command.go:
```diff
@@ -325,21 +383,73 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
+	return reply, externalIntents, nil
+}
+
+// intersectIntent takes an intent and a descriptor. It then splits the
```
This feels premature. Why not just treat a range of intents which exceeds the bounds of the range as non-local and send it through the distributed sender, which already implements this kind of logic.

spencerkimball replied:
On second thought, this is fine.

spencerkimball commented on storage/replica_command.go:
```diff
@@ -229,43 +230,63 @@ func (r *Replica) Scan(batch engine.Engine, args proto.ScanRequest) (proto.ScanR
-			return reply, proto.NewTransactionStatusError(existTxn, fmt.Sprintf("epoch regression: %d", txn.Epoch))
-		} else if txn.Epoch == existTxn.Epoch && existTxn.Timestamp.Less(txn.OrigTimestamp) {
+		if reply.Txn.Status == proto.COMMITTED {
+			return reply, nil, proto.NewTransactionStatusError(reply.Txn, "already committed")
```
In all of these error cases, don't we want to still resolve the intents? I think maybe you need another helper method for `EndTransaction` which settles the transaction status and then `EndTransaction` can resolve intents based on the final Transaction status.

tschottdorf replied:
if it's already committed, someone already sent an `EndTransaction` with a list of intents successfully, so all the machinery has run previously. For the regression errors, well they should simply never happen in practice. You can't actually ever resolve a non-local intent in that case until the transaction times out, at least not currently. So in those cases you'd want to limit yourself to local intents, which is exactly the part that we don't handle yet - returning a batch on error. I'm going to open an issue on this, it won't be hard to improve.

tschottdorf replied:
#1989

spencerkimball commented on storage/replica_command.go:
```diff
@@ -294,27 +312,67 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
+	// Resolve any explicit intents. All that are local to this range get
+	// resolved synchronously in the same batch. The remainder is collected
+	// and handed off to asynchronous processing.
+	var externalIntents []proto.Intent
```
I'd relocate the externalIntents var down above the for loop for a little more clarity.

tschottdorf replied:
done.

spencerkimball commented on storage/replica_command.go:
```diff
@@ -294,27 +312,67 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
+			// For intent ranges, cut into parts inside and outside our key
+			// range. Resolve locally inside, delegate the rest. In particular,
+			// an intent range for range-local data is correctly considered local.
+			// TODO(tschottdorf): This will panic for ranges which contain both
```
Why not add a test to storage/store.go in verifyKeys to make sure there's no cross over between non-local and local keyspace. That should be disallowed for sure.

tschottdorf replied:
done.

tschottdorf commented:
I'm ready to merge but going to wait for @bdarnell explaining the one comment I didn't understand.

tamird commented on storage/replica.go:
```diff
@@ -1145,6 +1150,103 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+	if !r.rm.Stopper().RunAsyncTask(action) {
+		// As with local intents, try async to not keep the caller waiting, but
+		// when draining just go ahead and do it synchronously. See #1684.
+		action()
```
this line appears to have resulted in a deadlock: https://circleci.com/gh/cockroachdb/cockroach/6250

tamird replied:
maybe not. https://github.com/cockroachdb/cockroach/issues/2206
